bindings/c/test/fdb.hpp

```diff
@@ -0,0 +1,543 @@
+	}
+}
+
+namespace network {
+
+inline Error set_option_nothrow(FDBNetworkOption option, BytesRef str) noexcept {
+	return Error(native::fdb_network_set_option(option, str.data(), intsize(str)));
+}
+
+inline Error set_option_nothrow(FDBNetworkOption option, int64_t value) noexcept {
+	return Error(native::fdb_network_set_option(
+	    option, reinterpret_cast<const uint8_t*>(&value), static_cast<int>(sizeof(value))));
+}
+
+inline void set_option(FDBNetworkOption option, BytesRef str) {
+	if (auto err = set_option_nothrow(option, str)) {
+		throw_error(fmt::format("ERROR: fdb_network_set_option({}): ",
+		                        static_cast<std::underlying_type_t<FDBNetworkOption>>(option)),
+		            err);
+	}
+}
+
+inline void set_option(FDBNetworkOption option, int64_t value) {
+	if (auto err = set_option_nothrow(option, value)) {
+		throw_error(fmt::format("ERROR: fdb_network_set_option({}, {}): ",
+		                        static_cast<std::underlying_type_t<FDBNetworkOption>>(option),
+		                        value),
+		            err);
+	}
+}
+
+inline Error setup_nothrow() noexcept {
+	return Error(native::fdb_setup_network());
+}
+
+inline void setup() {
+	if (auto err = setup_nothrow())
+		throw_error("ERROR: fdb_network_setup(): ", err);
+}
+
+inline Error run() {
+	return Error(native::fdb_run_network());
+}
+
+inline Error stop() {
+	return Error(native::fdb_stop_network());
+}
+
+} // namespace network
+
+class TX;
+class Database;
+
+class Result {
+	friend class TX;
+	std::shared_ptr<native::FDBResult> r;
+
+	Result(native::FDBResult* result) {
+		if (result)
+			r = std::shared_ptr<native::FDBResult>(result, &native::fdb_result_destroy);
+	}
+
+public:
+	using KeyValueArray = future_var::KeyValueArray::type;
+
+	Error get_keyvalue_array_nothrow(KeyValueArray& out) const noexcept {
+		auto out_more_native = native::fdb_bool_t{};
+		auto& [out_kv, out_count, out_more] = out;
+		auto err_raw = native::fdb_result_get_keyvalue_array(r.get(), &out_kv, &out_count, &out_more_native);
+		out_more = out_more_native != 0;
+		return Error(err_raw);
+	}
+
+	KeyValueArray get_keyvalue_array() const {
+		auto ret = KeyValueArray{};
+		if (auto err = get_keyvalue_array_nothrow(ret))
+			throw_error("ERROR: result_get_keyvalue_array(): ", err);
+		return ret;
+	}
+};
+
+class Future {
+protected:
+	friend class TX;
+	std::shared_ptr<native::FDBFuture> f;
+
+	Future(native::FDBFuture* future) {
+		if (future)
+			f = std::shared_ptr<native::FDBFuture>(future, &native::fdb_future_destroy);
+	}
+
+	// wrap any capturing lambda as callback passable to fdb_future_set_callback().
+	// destroy after invocation.
+	template <class Fn>
+	static void callback(native::FDBFuture*, void* param) {
+		auto fp = static_cast<Fn*>(param);
+		try {
+			(*fp)();
+		} catch (const std::exception& e) {
+			fmt::print(stderr, "ERROR: Exception thrown in user callback: {}", e.what());
+		}
+		delete fp;
```

Comment:
Not sure what you mean by this. If a future is canceled, it should result in a callback with an erroneous future as argument, and it will be deleted:
https://forums.foundationdb.org/t/does-fdb-future-set-callback-guarantee-exactly-once-execution/1498/3